### PR TITLE
Adding region constraints to pysensors PR

### PR DESCRIPTION
**Pull request by niharika2999**
This PR adds an optimizer GQR to pysensors for handling constrained optimization. 
The types of constraints it can handle are: 

- Region constrained max 
- Region constrained exact 
- Predetermined 

The functions for these constraints are added to utils.

### CODE REVIEW

**Review by joshua-cogliati-inl — COMMENTED** [review]
Just some comments after I looked at this PR.

File: pysensors/basis/_custom.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    U: The custom basis matrix
+
+    Attributes
+    ----------
+    basis_matrix_ : numpy ndarray, shape (n_features, n_basis_modes)
+        The top n_basis_modes left singular vectors of the training data.
+
+    """
+
+    def __init__(self, U, n_basis_modes=10, **kwargs):
```
[the remark this answers is not in the document]
Comment (niharika2999) [reply pysensors/basis/_custom.py:29]: It is not necessarily needed but we have included it for uniformity. Adding documentation for it.

File: examples/cost_constrained_qr.ipynb
[the remark this answers is not in the document]
Comment (niharika2999) [reply examples/cost_constrained_qr.ipynb:1]: No, this file needs to be added again. Thank you so much for catching that.

File: pysensors/basis/_custom.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (joshua-cogliati-inl) [reply pysensors/basis/_custom.py:29]: Thank you.

File: examples/cost_constrained_qr.ipynb
[the remark this answers is not in the document]
Comment (joshua-cogliati-inl) [reply examples/cost_constrained_qr.ipynb:1]: You are welcome, and this is now fixed.

**Pull request merged**